src/zenml/zen_server/request_management.py

```diff
@@ -0,0 +1,497 @@
+        method = self.request.method
+        return f"{method} {url_path} from {client_ip}"
+
+    @property
+    def log_duration(self) -> str:
+        """Get the duration of the request.
+
+        Returns:
+            The duration of the request.
+        """
+        current_time = utc_now()
+        duration = (current_time - self.received_at).total_seconds() * 1000
+        return f"{duration:.2f}ms"
+
+    @property
+    def is_cacheable(self) -> bool:
+        """Check if the request is cacheable.
+
+        Returns:
+            Whether the request is cacheable.
+        """
+        # Only cache requests that are authenticated and are part of a
+        # transaction.
+        return (
+            self.auth_context is not None and self.transaction_id is not None
+        )
+
+
+class RequestRecord:
+    """A record of an in-flight or cached request."""
+
+    future: asyncio.Future[Any]
+    request_context: RequestContext
+
+    def __init__(
+        self, future: asyncio.Future[Any], request_context: RequestContext
+    ) -> None:
+        """Initialize the request record.
+
+        Args:
+            future: The future of the request.
+            request_context: The request context.
+        """
+        self.future = future
+        self.request_context = request_context
+        self.completed = False
+
+    def set_result(self, result: Any) -> None:
+        """Set the result of the request.
+
+        Args:
+            result: The result of the request.
+        """
+        self.future.set_result(result)
+        self.completed = True
+
+    def set_exception(self, exception: Exception) -> None:
+        """Set the exception of the request.
+
+        Args:
+            exception: The exception of the request.
+        """
+        self.future.set_exception(exception)
+        self.completed = True
+
+
+class RequestManager:
+    """A manager for requests.
+
+    This class is used to manage requests by caching the results of requests
+    that have already been executed.
+    """
+
+    def __init__(
+        self,
+        transaction_ttl: int,
+        request_timeout: float,
+        deduplicate: bool = True,
+    ) -> None:
+        """Initialize the request manager.
+
+        Args:
+            transaction_ttl: The time to live for cached transactions. Comes
+                into effect after a request is completed.
+            request_timeout: The timeout for requests. If a request takes longer
+                than this, a 429 error will be returned to the client to slow
+                down the request rate.
+            deduplicate: Whether to deduplicate requests.
+        """
+        self.deduplicate = deduplicate
+        self.transactions: Dict[UUID, RequestRecord] = dict()
+        self.lock = asyncio.Lock()
+        self.transaction_ttl = transaction_ttl
+        self.request_timeout = request_timeout
+
+        self.request_contexts: ContextVar[Optional[RequestContext]] = (
```

Comment:
This is only one right, so `request_context` makes more sense?

Reply:
It's a matter of opinion. These are multiple variables, one for each coroutine or thread being executed. It's only one if you look at it from the POV of a single coroutine/thread.